packages/backend-common/src/hot.ts

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * This function allows devs to cleanup
+ * ongoing effects when module gets hot-reloaded
+ * Useful for cleaning intervals, timers, requests etc
+ * @example
+ * ```ts
+ * const intervalId = setInterval(doStuff, 1000);
+ * useHotCleanup(module, () => clearInterval(intervalId));
+ * ```
+ * @param _module Reference to the current module where you invoke the fn
+ * @param cancelEffect Fn that cleans up the ongoing effects
+ */
+export function useHotCleanup(_module: NodeModule, cancelEffect: () => void) {
+  if (_module.hot) {
+    _module.hot.addDisposeHandler(() => {
+      cancelEffect();
+    });
+  }
+}
+
+/**
+ * This function allows devs to preserve
+ * some value between hot-reloads.
+ * Useful for stateful parts of the backend
+ * @example
+ * ```ts
+ * const db = useHotMemoize(module, () => createDB(dbParams));
+ * ```
+ * @param _module Reference to the current module where you invoke the fn
+ * @param valueFactory Fn that returns the value you want to memoize
+ * @warning Don't use inside conditionals or loops,
+ * same rules as for hooks apply (https://reactjs.org/docs/hooks-rules.html)
+ */
+export function useHotMemoize<T>(
+  _module: NodeModule,
+  valueFactory: () => T,
+): T {
+  const CURRENT_HOT_MEMOIZE_INDEX_KEY = 'backstage.io/hmr-memoize-key';
```

Comment:
Would it be better to use a symbol here instead? Assuming none of this is serialized

Reply:
Yes, good idea!

Reply:
Realized that might not be possible because we'd get a new symbol each time. If we could use a symbol here we could've just closed over values and not used the module.hot.data to begin with.